File: android/dimina/src/main/kotlin/com/didi/dimina/ui/view/CanvasImageLoader.kt
package com.didi.dimina.ui.view

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.util.Log
import okhttp3.OkHttpClient
import okhttp3.Request
import org.json.JSONObject
import java.util.concurrent.ConcurrentHashMap

/**
 * Image download+decode helper for native canvas.
 * Handles imageSetSrc ops: downloads images via URL, decodes to RGBA pixels,
 * and uploads to the native GL canvas via NativeCanvasGLView.nativeUploadImage.
 */
object CanvasImageLoader {
    private const val TAG = "CanvasImageLoader"

    /** Registry of active NativeCanvasGLView instances by nodeId */
    private val glViews = ConcurrentHashMap<String, NativeCanvasGLView>()

    private val httpClient = OkHttpClient()

    fun registerView(nodeId: String, view: NativeCanvasGLView) {
        glViews[nodeId] = view
    }

    fun unregisterView(nodeId: String) {
        glViews.remove(nodeId)
    }

    /**
     * Handle an imageSetSrc request from native canvas_bindings.
     * Called on the main thread. Downloads the image on a background thread,
     * decodes to RGBA, and calls nativeUploadImage.
     *
     * @param instanceId the QuickJSEngine instance
     * @param json JSON with {nodeId, imageId, src, onload, onerror}
     */
    fun handleImageSetSrc(instanceId: Int, json: String) {
        try {
            val obj = JSONObject(json)
            val nodeId = obj.optString("nodeId")
            val imageId = obj.optString("imageId")
            val src = obj.optString("src")
            val onload = obj.optString("onload", "")
            val onerror = obj.optString("onerror", "")

            if (nodeId.isEmpty() || imageId.isEmpty() || src.isEmpty()) {
                Log.e(TAG, "handleImageSetSrc: missing fields in json")
                return
            }

            Log.d(TAG, "handleImageSetSrc: nodeId=$nodeId imageId=$imageId src=$src")

            // Download and decode on a background thread
            Thread {
                try {
                    val bitmap = downloadAndDecode(src)
                    if (bitmap != null) {
                        val rgba = bitmapToRGBA(bitmap)
                        val width = bitmap.width
                        val height = bitmap.height
                        bitmap.recycle()

                        // Upload to native canvas
                        val view = glViews[nodeId]
                        if (view != null) {
                            view.uploadImage(imageId, rgba, width, height, onload)
                            Log.d(TAG, "handleImageSetSrc: uploaded imageId=$imageId ${width}x${height}")
                        } else {
                            Log.e(TAG, "handleImageSetSrc: no GLView for nodeId=$nodeId")
                        }
                    } else {
                        Log.e(TAG, "handleImageSetSrc: decode failed for src=$src")
                        // TODO: invoke onerror callback if needed
                    }
                } catch (e: Exception) {
                    Log.e(TAG, "handleImageSetSrc: error downloading/decoding", e)
                }
            }.start()
        } catch (e: Exception) {
            Log.e(TAG, "handleImageSetSrc: JSON parse error", e)
        }
    }

    private fun downloadAndDecode(src: String): Bitmap? {
        return try {
            if (src.startsWith("http://") || src.startsWith("https://")) {
                // Network image
                val request = Request.Builder().url(src).build()
                val response = httpClient.newCall(request).execute()
                response.use {
                    if (it.isSuccessful) {
                        val bytes = it.body?.bytes()
                        if (bytes != null) {
                            BitmapFactory.decodeByteArray(bytes, 0, bytes.size)
                        } else null
                    } else null
                }
            } else {
                // Local file
                BitmapFactory.decodeFile(src)
            }
        } catch (e: Exception) {
            Log.e(TAG, "downloadAndDecode: failed for src=$src", e)
            null
        }
    }

    private fun bitmapToRGBA(bitmap: Bitmap): ByteArray {
        // Ensure ARGB_8888 format
        val argbBitmap = if (bitmap.config != Bitmap.Config.ARGB_8888) {
            bitmap.copy(Bitmap.Config.ARGB_8888, false)
        } else {
            bitmap
        }

        val width = argbBitmap.width
        val height = argbBitmap.height

        // Extract pixel data and convert ARGB → RGBA
        val pixels = IntArray(width * height)
        argbBitmap.getPixels(pixels, 0, width, 0, 0, width, height)

        if (argbBitmap !== bitmap) {
            argbBitmap.recycle()
        }

        val rgba = ByteArray(width * height * 4)
        for (i in pixels.indices) {
            val pixel = pixels[i]
            val offset = i * 4
            rgba[offset] = ((pixel shr 16) and 0xFF).toByte()     // R
            rgba[offset + 1] = ((pixel shr 8) and 0xFF).toByte()  // G
            rgba[offset + 2] = (pixel and 0xFF).toByte()           // B
            rgba[offset + 3] = ((pixel shr 24) and 0xFF).toByte() // A
        }
        return rgba
    }
}

File: iOS/dimina/DiminaKit/Container/Api/UI/CanvasImageLoader.swift
//
//  CanvasImageLoader.swift
//  dimina
//
//  Image download + decode helper for native GL canvas.
//  Handles imageSetSrc ops: downloads images via URL, decodes to RGBA pixels,
//  and uploads to the native GL canvas. Equivalent of Android's CanvasImageLoader.kt.
//

import Foundation
import UIKit

class CanvasImageLoader {
    static let shared = CanvasImageLoader()

    /// Registry of active NativeCanvasGLView instances by nodeId
    private var glViews: [String: NativeCanvasGLView] = [:]

    /// Weak reference to engine for scheduling JS callbacks
    private weak var engine: DMPEngine?

    private init() {}

    func setEngine(_ engine: DMPEngine) {
        self.engine = engine
    }

    func registerView(nodeId: String, view: NativeCanvasGLView) {
        glViews[nodeId] = view
    }

    func unregisterView(nodeId: String) {
        glViews.removeValue(forKey: nodeId)
    }

    /// Handle an imageSetSrc request from DMPEngineCanvas.
    /// Downloads the image on a background queue, decodes to RGBA pixels,
    /// and uploads to the GL canvas via dm_canvas_load_image_rgba.
    func handleImageSetSrc(engine: DMPEngine, nodeId: String, imageId: String,
                           src: String, onload: String, onerror: String) {
        DispatchQueue.global(qos: .userInitiated).async { [weak self] in
            guard let self = self else { return }

            self.downloadAndDecode(src: src) { result in
                switch result {
                case .success(let (rgba, width, height)):
                    // Upload on the JS thread (which owns the EGL context)
                    engine.performOnJSThreadInternal { [weak self] in
                        guard let self = self, let view = self.glViews[nodeId] else {
                            DMPLogger.debug("CanvasImageLoader: no GLView for nodeId=\(nodeId)")
                            return
                        }

                        // Direct upload via C API on JS/GL thread
                        if let canvas = view.getCanvasHandle() {
                            rgba.withUnsafeBytes { rawBuf in
                                guard let ptr = rawBuf.baseAddress?.assumingMemoryBound(to: UInt8.self) else { return }
                                let rc = dm_canvas_load_image_rgba(canvas, imageId, Int32(width), Int32(height), ptr)
                                if rc == 0 && !onload.isEmpty {
                                    let argsJson = "{\"width\":\(width),\"height\":\(height)}"
                                    self.invokeCallback(engine: engine, callbackId: onload, argsJson: argsJson)
                                    DMPLogger.debug("CanvasImageLoader: uploaded \(imageId) \(width)x\(height)")
                                } else if rc != 0 {
                                    DMPLogger.debug("CanvasImageLoader: upload failed \(imageId)")
                                    if !onerror.isEmpty {
                                        self.invokeCallback(engine: engine, callbackId: onerror,
                                                          argsJson: "{\"errMsg\":\"createImage:fail upload error\"}")
                                    }
                                }
                            }
                        }
                    }

                case .failure(let error):
                    DMPLogger.debug("CanvasImageLoader: decode failed src=\(src) error=\(error)")
                    if !onerror.isEmpty {
                        engine.enqueueScript("""
                            (function(){
                                var cb=globalThis.__dimina_callback_registry;
                                if(cb&&cb.invoke)cb.invoke('\(onerror)',{"errMsg":"createImage:fail \(error.localizedDescription)"});
                            })()
                            """)
                    }
                }
            }
        }
    }

    // MARK: - Private

    private func downloadAndDecode(src: String, completion: @escaping (Result<(Data, Int, Int), Error>) -> Void) {
        if src.hasPrefix("http://") || src.hasPrefix("https://") {
            guard let url = URL(string: src) else {
                completion(.failure(NSError(domain: "CanvasImageLoader", code: -1,
                                           userInfo: [NSLocalizedDescriptionKey: "Invalid URL"])))
                return
            }

            URLSession.shared.dataTask(with: url) { data, _, error in
                if let error = error {
                    completion(.failure(error))
                    return
                }
                guard let data = data, let image = UIImage(data: data) else {
                    completion(.failure(NSError(domain: "CanvasImageLoader", code: -2,
                                               userInfo: [NSLocalizedDescriptionKey: "Decode failed"])))
                    return
                }
                if let rgba = self.imageToRGBA(image) {
                    completion(.success(rgba))
                } else {
                    completion(.failure(NSError(domain: "CanvasImageLoader", code: -3,
                                               userInfo: [NSLocalizedDescriptionKey: "RGBA conversion failed"])))
                }
            }.resume()
        } else {
            // Local file
            guard let image = UIImage(contentsOfFile: src) else {
                completion(.failure(NSError(domain: "CanvasImageLoader", code: -4,
                                           userInfo: [NSLocalizedDescriptionKey: "File not found: \(src)"])))
                return
            }
            if let rgba = imageToRGBA(image) {
                completion(.success(rgba))
            } else {
                completion(.failure(NSError(domain: "CanvasImageLoader", code: -5,
                                           userInfo: [NSLocalizedDescriptionKey: "RGBA conversion failed"])))
            }
        }
    }

    /// Convert UIImage to raw RGBA pixel data.
    private func imageToRGBA(_ image: UIImage) -> (Data, Int, Int)? {
        guard let cgImage = image.cgImage else { return nil }

        let width = cgImage.width
        let height = cgImage.height
        let bytesPerRow = width * 4

        var rgbaData = Data(count: height * bytesPerRow)
        let colorSpace = CGColorSpaceCreateDeviceRGB()

        guard let context = rgbaData.withUnsafeMutableBytes({ rawBuf -> CGContext? in
            CGContext(data: rawBuf.baseAddress, width: width, height: height,
                      bitsPerComponent: 8, bytesPerRow: bytesPerRow,
                      space: colorSpace,
                      bitmapInfo: CGImageAlphaInfo.premultipliedLast.rawValue | CGBitmapInfo.byteOrder32Big.rawValue)
        }) else { return nil }

        context.draw(cgImage, in: CGRect(x: 0, y: 0, width: width, height: height))

        return (rgbaData, width, height)
    }

    /// Invoke a JS callback via the dimina callback registry
    private func invokeCallback(engine: DMPEngine, callbackId: String, argsJson: String) {
        let script = """
            (function(){
                var cb=globalThis.__dimina_callback_registry;
                if(cb&&cb.invoke)cb.invoke('\(callbackId)',\(argsJson));
            })()
            """
        engine.enqueueScript(script)
    }
}
